core/src/main/java/com/linecorp/armeria/internal/logging/ContentPreviewingUtil.java

```diff
@@ -71,19 +76,40 @@ protected HttpObject filter(HttpObject obj) {
 
             @Override
             protected void beforeComplete(Subscriber<? super HttpObject> subscriber) {
-                logBuilder.requestContentPreview(requestContentPreviewer.produce());
+                produceRequestContentPreview(requestContentPreviewer);
             }
 
             @Override
             protected Throwable beforeError(Subscriber<? super HttpObject> subscriber,
                                             Throwable cause) {
-                // Call produce() to release the resources in the previewer. Consider adding close() method.
-                requestContentPreviewer.produce();
+                try {
+                    // Call produce() to release the resources in the previewer. Consider adding close() method.
+                    requestContentPreviewer.produce();
+                } catch (Exception e) {
+                    logger.warn("Unexpected exception while producing the request content preview. " +
+                                "previewer: {}", requestContentPreviewer, e);
+                }
 
                 // Set null to make it sure the log is complete.
                 logBuilder.requestContentPreview(null);
                 return cause;
             }
+
+            @Override
+            protected void beforeCancel(Subscription subscription) {
+                produceRequestContentPreview(requestContentPreviewer);
+            }
+
+            private void produceRequestContentPreview(ContentPreviewer requestContentPreviewer) {
```

Comment:
This method could be called twice if a `StreamMessage` subscribed with `NOTIFY_CANCELLATION` is canceled.